File: Problems/0001_0099/0067_Add_Binary/Project_CS/Add_Binary.cs
using System;
using System.Text;

public class Solution
{
    public string AddBinary(string a, string b)
    {
        // 72ms - 85ms 
        string ans = "";
        int carry = 0;
        int i = a.Length - 1;
        int j = b.Length - 1;
        while ((i >= 0) || (j >= 0) || carry != 0) {
            if (i >= 0) {
                carry += a[i--] - '0';
            }
            if (j >= 0) {
                carry += b[j--] - '0';
            }
            ans = (char)(carry % 2 + '0') + ans;
            carry /= 2;
        }
        return ans;
    }

    private long ToDecimal(string workStr)
    {
        long val = 0, n = 1;
        
        for (int i = workStr.Length - 1; i >= 0; --i) {
            val += (workStr[i] - '0')*n;
            n *= 2;
        }
        return (val);
    }
    
    public void Main(string args)
    {
        string arg_str = args.Replace("[[","").Replace("]]","").Replace("\"", "").Trim();
        string[] flds = arg_str.Split(new string[] {"],["}, StringSplitOptions.None);
        string a = flds[0];
        string b = flds[1];
        Console.WriteLine("a = " + ToDecimal(a).ToString());
        Console.WriteLine("b = " + ToDecimal(b).ToString());

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();

        sw.Start();

        string result = AddBinary(a, b);
        
        sw.Stop();

        Console.WriteLine("Result(bin) = \"" + result + "\", Result(dec) = " + ToDecimal(result) );
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0001_0099/0067_Add_Binary/Project_Python3/Add_Binary.py
import os
import sys
import time
from typing import List, Dict, Tuple

class Solution:
    def addBinary_1liner(self, a: str, b: str) -> str:
        # 43ms - 56ms
        return bin(int(a, 2) + int(b, 2))[2:]

    def addBinary(self, a: str, b: str) -> str:
        # 53ms - 62ms
        s = ""
        carry = 0
        i, j = len(a) - 1, len(b) - 1
        while i >= 0 or j >= 0 or carry:
            if i >= 0:
                carry += int(a[i])
                i -= 1
            if j >= 0:
                carry += int(b[j])
                j -= 1
            s = str(carry % 2) + s
            carry //= 2
        return s

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python {0} <testdata.txt>".format(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("{0} not found...".format(argv[1]))
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = {0}".format(temp))
        loop_main(temp)
    #   print("Hit Return to continue...")
    #   input()

def loop_main(temp):
    flds = temp.replace("\"","").replace("[","").replace("]","").rstrip().split(",")
    a, b = flds[0], flds[1]
    print("a = \"{0}\", b = \"{1}\"".format(a, b))

    sl = Solution()
    time0 = time.time()

    result = sl.addBinary(a, b)

    time1 = time.time()

    print("result = \"{0}\"".format(result))
    print("Execute time ... : {0:f}[s]\n".format(time1 - time0))

if __name__ == "__main__":
    main()
